### ISSUE TITLE
Question: Cross-Entropy and variable sized logits from dynamic_decode

### ISSUE THREAD

**Issue by sleighsoft** [tensorflow/nmt#2]
I have a very similar setup also using the new v1.2 APIs.
When I run the encoder+decoer with features of size e.g. `[2, 23, 100]` ([batch, sentence_length, source_vocabulary_size]) it will sometimes return a differntly sized Tensor e.g. `[2, 16, 100]` from `dynamic_decode`. I know this is intended, as the target sentence can have a different length from the input sentence.

What I do not understand is the following:
Let's assume I have a label tensor of shape `[2, 30]` and `dynamic_decode` outputs `[2, 16, 100]` how do I compute the crossentropy like it is done here:
```
target_output = self.iterator.target_output
if self.time_major:
  target_output = tf.transpose(target_output)
max_time = self.get_max_time(target_output)
crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(
    labels=target_output, logits=logits)
target_weights = tf.sequence_mask(
    self.iterator.target_sequence_length, max_time, dtype=logits.dtype)
if self.time_major:
  target_weights = tf.transpose(target_weights)

loss = tf.reduce_sum(
    crossent * target_weights) / tf.to_float(self.batch_size)
return loss
```
https://github.com/tensorflow/nmt/blob/master/nmt/model.py#L414

When the shapes of `logits` and `labels` do not match. I currently pad/trim my logits in that case but I don't see that happening in the code here.
Could that point to a bug in my implementation or do I misunderstand something?
I know this might be more suitable for stackoverflow but I feel like I miss an important part of the model that is presented here.
I hope someone can clarify that for me.

**Comment by ebrevdo** [comment]
It sounds like you're using a decoder which is not a training decoder (i.e., you're using a beam search or inference decoder); these can finish "early".

These aren't meant to train with - only perform inference (though we're thinking about ways to expand the beam search decoder to work in a training setting; it's not done yet).

If you use a Training decoder, or i.e., a scheduled sampling (embedding) decoder, then the number of time steps out should always match the input to the decoder.

Without seeing some example code, that's my best guess.

**Comment by okuchaiev** [comment]
@sleighsoft , you can slice your logits and targets to the common length. 

Something like this (assuming batch-major):

`current_ts = tf.to_int32(tf.minimum(tf.shape(target_output)[1], tf.shape(logits)[1]))`
`target_sequence = tf.slice(target_sequence, begin=[0, 0], size=[-1, current_ts])`
`mask_ = tf.sequence_mask(lengths=target_sequence_length,  maxlen=current_ts, dtype=logits.dtype)`
 `logits = tf.slice(logits, begin=[0,0,0], size=[-1, current_ts, -1])`

**Comment by sleighsoft** [comment]
@okuchaiev I am currently doing that but I am wondering why it is not done in the nmt code.
@ebrevdo I believe I am doing what you propose, maybe you can find what's wrong with my implementation. Thanks for looking into it :)

This is part of my code.
Sorry for it being a bit messy but I started with it when TF1.2 documentation was scarce and I am trying out a lot of things and added some lines here and there over time.
```python
# Cell setup
def lstm_cell():
  return rnn.LSTMCell(self.cell_size)

if self.dropout is not None and mode == learn.ModeKeys.TRAIN:
  def lstm_dropout_cell():
    return rnn.DropoutWrapper(
        rnn.LSTMCell(self.cell_size),
        input_keep_prob=self.dropout)
  lstm_cell = lstm_dropout_cell

def cell():
  if self.layers > 1:
    cell = rnn.MultiRNNCell(
        [lstm_cell() for _ in range(self.layers)])
    if self.dropout is not None and mode == learn.ModeKeys.TRAIN:
      cell = rnn.DropoutWrapper(cell, output_keep_prob=self.dropout)
  else:
    cell = lstm_cell()
  return cell

# Encoder
# First batch: features = [2 31 25]
# Second batch: features = [2 31 25]
with tf.name_scope('encoder'):
  encoder_cell = cell()
  _, encoder_state = tf.nn.dynamic_rnn(
      cell=encoder_cell,
      inputs=features,
      sequence_length=params['enc_seq_len'],
      dtype=tf.float32)

# Decoder
with tf.variable_scope('decoder') as decoder_scope:
  decoder_cell = cell()
  embedding = tf.get_variable(
      "decoder_embedding", [self.decoder_vocabulary_size, self.cell_size])
  batch_size = tf.shape(features)[0]
  if mode != learn.ModeKeys.INFER:
    # Train helper for eval and train
    decoder_emb_inp = tf.nn.embedding_lookup(embedding, labels)
    # First batch: decoder_emb_inp= [2 30 25]
    # Second batch: decoder_emb_inp= [2 30 25]
    helper = seq2seq.TrainingHelper(
        decoder_emb_inp, params['dec_seq_len'], time_major=False)
  else:
    # Greedy for inference
    helper = seq2seq.GreedyEmbeddingHelper(
        embedding=embedding,
        start_tokens=tf.tile([GO_SYMBOL], [batch_size]),
        end_token=END_SYMBOL)

  decoder = seq2seq.BasicDecoder(
      cell=decoder_cell,
      helper=helper,
      initial_state=encoder_state,
      output_layer=Dense(self.decoder_vocabulary_size, use_bias=False,
                         name='output_projection'))

  decoding_length_factor = 2
  max_iterations = self.max_sequence_length * decoding_length_factor

  if mode != learn.ModeKeys.INFER:
    max_iterations = None
  decoder_tuple = seq2seq.dynamic_decode(
      decoder=decoder,
      output_time_major=False,
      swap_memory=True,
      maximum_iterations=max_iterations,
      scope=decoder_scope)

decoder_outputs = decoder_tuple[0]
# logits = [batch x seq_len x decoder_vocabulary_size]
logits = decoder_outputs.rnn_output
# predictions = [batch x seq_len]
predictions = decoder_outputs.sample_id

...
# Loss
weights = tf.sequence_mask(tf.to_int32(params['dec_seq_len']),
                           tf.to_int32(label_length),
                           dtype=logits.dtype)
# First batch: labels = [2 30], logits = [2 30 189]
# Second batch: labels = [2 30], logits = [2 14 189]
crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits)
loss = tf.reduce_sum(crossent * weights) / tf.to_float(tf.shape(features)[0])
```

I have annotated the Tensor shapes in code above. In the `#Loss` section I noted the "wrong" logit shape.

I use the `TrainingHelper` for `train` and `evaluation` .
The `GreedyEmbeddingHelper` for `inference`.

**Comment by sleighsoft** [comment]
@ebrevdo Your hint also helped my find the reason for the shape mismatch. I have a multi GPU setup where I split the batches of labels and features in half. What can happen is that I split my batch with lengths e.g. [30 10 14 10] into [30 10] and [14 10]. This results in one `dynamic_decode` outputting [2 30 189] and the other one [2 14 189] as the maximum `params['dec_seq_len']` is shorther in one part of the batch.

**Comment by sleighsoft** [comment]
Maybe you'll find another bug in my code as it tends to predict the same word over and over again even though loss converges to 0.

**Comment by sleighsoft** [comment]
So, I also found the solution for my second problem. I trained without at `start of sentence` symbol in the `TrainingHelper` but used the `sos` symbol for `inference` thus it learned to translate `x -> x`.

**Comment by ebrevdo** [comment]
Ah ok great.  I wish there was a part of the API that encouraged feeding
SOS tokens even to TrainingHelper, but I think it might be too
restrictive.  Perhaps we can add that in the docstring for TrainingHelper?
Feel free to open an issue for this on tensorflow/tensorflow project issues
(or send a PR there if you want)

**Comment by sleighsoft** [comment]
If it is not too much of a hassle I will do a PR on tensorflow/tensorflow.
Thanks for the help!

**Comment by okuchaiev** [comment]
@sleighsoft 
I think I might be having similar issue. 
When using BasicDecoder+TrainingHelper for training the loss goes down really quickly. And I can learn simple tasks (like reversing strings). But on real task, inference get stuck at same word even thought training loss was low and I got very good samples during training. 
What do you mean by "learned to translate x -> x" ? Can you learn to translate De->En ? or any other reasonably big task?

**Comment by okuchaiev** [comment]
@ebrevdo Why does the GreedyEmbedding helper actually require GO_SYMBOL?
Shouldn't decoder actually learn to start with this? E.g. given encoder/attention outputs and zero state, the decoder should learn to predict GO_SYMBOL

**Comment by ebrevdo** [comment]
This is the standard approach I've seen; people feed a go symbol to start
with.  What if you don't feed zero state, but a bottleneck state from the
output of the encoder rnn instead?

On Fri, Jul 14, 2017 at 11:53 AM, Oleksii Kuchaiev <notifications@github.com

**Comment by tocab** [comment]
@ebrevdo 

> If you use a Training decoder, or i.e., a scheduled sampling (embedding) decoder, then the number of time steps out should always match the input to the decoder.

I don't think that this is right. I spend the whole day yesterday to figure out why my encoder decoder system doesn't work and it came out that the decoder stops when it has processed the sequence with the largest sequence length. E.g. when by batched Decoder inputs with batch size 64 and sequence length 50 has shape [64,50] , the decoder output can have a shape like [64,48]. I give the decoder also a sequence length vector, where in this case the max sequence length is 48. So the decoder seems to stop decoding when it has reached the max sequence length.

It would be very helpful if this could be documented.

**Comment by gabox8888** [comment]
@tocab  

I think I am experiencing a similar error, would you mind sharing how you solved it?

**Comment by ebrevdo** [comment]
Ah, I think you're right.  The documentation needs to be added to the
constructor docstrings for the TrainingHelper and possibly the
ScheduledSamplingHelper.  Do you want to send a PR with these small
documentation improvements?

**Comment by gabox8888** [comment]
@ebrevdo 

I believe I am using the api correctly, yet i keep getting a shape mismatch error.

`    max_target_sequence_length = tf.reduce_max(tf.reduce_sum(self.net.mask, 1), name='max_target_len')
            masks = tf.sequence_mask(tf.reduce_sum(self.net.mask, 1), max_target_sequence_length, dtype=tf.float32, name='masks')

            logit_length = tf.shape(self.net.current_V)[1]
            logit_length = tf.Print(logit_length , [logit_length ], message="This is logits length: ")
            pad_size = 130 - logit_length
            logits = tf.pad(self.net.current_V, [[0, 0], [0, pad_size], [0, 0]])
            new_l = tf.shape(logits)
            new_l = tf.Print(new_l , [new_l ], message="This is logits: ")
            loss = tf.contrib.seq2seq.sequence_loss(logits,self.net.labels,masks)`

This is my loss function, and this is how i generate the logits

`sequence_length = tf.constant(130, shape=[14])#tf.reduce_max(tf.reduce_sum(self.mask, axis=1)) 
            sequence_length = tf.Print(sequence_length, [sequence_length], message="This is a: ")

            dec_input = process_decoder_input(self.labels,14)
            dec_embeddings = tf.Variable(tf.random_uniform([5004, 512]))
            dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
            dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(number_units) for _ in range(2)])
            output_layer = layers_core.Dense(5004, use_bias=False)
            
            training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,sequence_length=sequence_length,time_major=False)
        
            training_decoder = tf.contrib.seq2seq.BasicDecoder(cell=dec_cell,helper=training_helper,initial_state=(self.current_V,self.current_V),output_layer=output_layer) 
            training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder,impute_finished=True,maximum_iterations=260)[0] 
            decoder_output = training_decoder_output.rnn_output`

Yet I keep getting that shape error. I am not sure how to fix it, any chance you could point me in the right direction ?

**Comment by sleighsoft** [comment]
Can you reformat your code so that all code is in a code block.
I had multiple errors in my code, to which one do you refer?

**Comment by vsuthichai** [comment]
I'm having a similar problem with the shaping error, and I thought I'd post my code here.
```
def lstm_model(args, np_embeddings, mode='train'):
    vocab_size = np_embeddings.shape[0]

    # Embeddings
    with tf.variable_scope('embeddings'):
        embeddings = tf.get_variable(name="embeddings", shape=np_embeddings.shape, initializer=tf.constant_initializer(np_embeddings), trainable=True)

    # Define placeholders
    with tf.variable_scope('placeholders'):
        seq_source_ids = tf.placeholder(tf.int32, shape=(None, args.max_seq_length), name="source_ids")
        seq_source_lengths = tf.placeholder(tf.int32, [None], name="sequence_source_lengths")
        seq_reference_ids = tf.placeholder(tf.int32, shape=(None, args.max_seq_length), name="reference_ids")
        seq_reference_lengths = tf.placeholder(tf.int32, [None], name="sequence_reference_lengths")
        seq_output_ids = tf.placeholder(tf.int32, shape=(None, args.max_seq_length), name="output_ids")

    # Encoder embeddings
    with tf.variable_scope('encoder_embeddings'):
        encoder_embedding = tf.nn.embedding_lookup(embeddings, seq_source_ids, name="encoder_embeddings")

    # Encoder
    with tf.variable_scope('encoder'):
        encoder_cell = tf.nn.rnn_cell.BasicLSTMCell(args.hidden_size)
        encoder_outputs, encoder_states = tf.nn.dynamic_rnn(cell=encoder_cell, inputs=encoder_embedding, sequence_length=seq_source_lengths, dtype=tf.float32)

    # Decoder embeddings
    with tf.variable_scope('decoder_embeddings'):
        decoder_embedding = tf.nn.embedding_lookup(embeddings, seq_reference_ids, name="decoder_embeddings")

    # Decoder
    with tf.variable_scope('decoder'):
        projection_layer = layers_core.Dense(vocab_size, use_bias=False)
        decoder_cell = tf.nn.rnn_cell.BasicLSTMCell(args.hidden_size)
        helper = tf.contrib.seq2seq.TrainingHelper(decoder_embedding, seq_reference_lengths)
        decoder = tf.contrib.seq2seq.BasicDecoder(decoder_cell, helper, encoder_states, output_layer=projection_layer)
        final_outputs, final_state, final_sequence_lengths = tf.contrib.seq2seq.dynamic_decode(decoder)
        #print(final_outputs)
        #print(final_state)
        #print(final_sequence_lengths.shape)
        logits = final_outputs.rnn_output

    # Loss
    with tf.variable_scope('loss'):
        pad = tf.fill((tf.shape(seq_output_ids)[0], args.max_seq_length), -1)
        boolean_mask = tf.not_equal(seq_output_ids, pad)
        mask = tf.cast(boolean_mask, tf.float32)
        #print(seq_output_ids.shape)
        #print(logits.shape)
        crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=seq_output_ids, logits=logits)
        train_loss = (tf.reduce_sum(crossent * mask) / args.batch_size)
```
This is the error I get:
```
InvalidArgumentError (see above for traceback): logits and labels must have the same first dimension, got logits shape [16,400002] and labels shape [60]
```

Any thoughts on how I might go about fixing this?

**Comment by sleighsoft** [comment]
What happens in `tf.nn.sparse_softmax_cross_entropy_with_logits` is the following. It computes the shapes like this:
Logits shape = `16 * 400002`
Labels shape = `60`

Therefore you have to make sure logits computes to the same shape as labels.

From quickly looking at your code I do not see any immediate errors. Just double and triple check your intermediate shapes.

**Comment by vsuthichai** [comment]
I've noticed that the output of `tf.contrib.seq2seq.dynamic_decode` will have `rnn_output` with shape [batch_size, max_sentence_ref_length across entire batch, num_classes] while the labels are the max sentence length with all the padding (this ends up being longer than the max_sentence_ref_length from `rnn_output`).  So it looks like I need to reshape my labels and remove the padding?

**Comment by sleighsoft** [comment]
The `sequence_length` parameter of `TrainingHelper` should be a vector containing the individual sequence lengths of all your labels within your batch, excluding the padding. Also, set `maximum_iterations` of dynamic decode to the maximum sequence length within each batch. This should fix your shape issues.

**Comment by vsuthichai** [comment]
@sleighsoft Would you be able to provide some guidance on completing my lstm decoder?  Here is what I have so far.

```
      with tf.variable_scope('decoder_embedding'):
          decoder_embedding = tf.nn.embedding_lookup(embeddings, seq_reference_ids, name="decoder_embedding")

        with tf.variable_scope('decoder'):
            decoder_cell = tf.nn.rnn_cell.BasicLSTMCell(args.hidden_size * 2)
            helper = tf.contrib.seq2seq.TrainingHelper(decoder_embedding, seq_reference_lengths)
            fc_layer = layers_core.Dense(vocab_size, use_bias=False)
            decoder = tf.contrib.seq2seq.BasicDecoder(decoder_cell, helper, joined_encoder_state, fc_layer)
            final_outputs, final_state, final_sequence_lengths = tf.contrib.seq2seq.dynamic_decode(decoder, swap_memory=True)
            logits = final_outputs.rnn_output
            predictions = final_outputs.sample_id
```

The loss function is as follows.  This does not work. `seq_output_ids` are the target ids.  Is there any padding or masking I need to do in order to get the shapes to match up?  Do you use a start and end token in the targets for your batch?  Any help would be greatly appreciated :)
```
            targets = tf.reshape(seq_output_ids, [-1])
            logits_flat = tf.reshape(logits, [-1, vocab_size])
            crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=targets, logits=logits_flat)
           train_loss = (tf.reduce_sum(crossent) / batch_size)
```
